python/rascaline/rascaline/utils/clebsch_gordan/_lambda_soap.py

```diff
@@ -0,0 +1,288 @@
+    operations,
+    torch_jit_export,
+    torch_jit_is_scripting,
+)
+from . import _coefficients, _utils
+from ._correlate_density import DensityCorrelations
+from ._correlate_tensors import CorrelateTensorWithDensity
+
+
+try:
+    import torch
+
+    HAS_TORCH = True
+except ImportError:
+    HAS_TORCH = False
+
+# ======================================================================
+# ===== Public API functions
+# ======================================================================
+
+
+class LambdaSoap(TorchModule):
+    """
+    Computes a Lambda-SOAP equivariant descriptor.
+    """
+
+    def __init__(
+        self,
+        spherical_expansion_hypers: dict,
+        atom_types: Optional[List[int]] = None,
+        angular_cutoff: Optional[int] = None,
+        selected_keys: Optional[Labels] = None,
+    ) -> None:
+
+        self._spherical_expansion_hypers = spherical_expansion_hypers
+        self._atom_types = atom_types
+        self._spherical_expansion_calc = SphericalExpansion(
+            **spherical_expansion_hypers
+        )
+
+        density_correlations_hypers = {
+            "max_angular": self._spherical_expansion_hypers["max_angular"] * 2,
+            "body_order": 3,
+            "angular_cutoff": angular_cutoff,
+            "selected_keys": selected_keys,
+            "match_keys": ["center_type"],
+            "skip_redundant": True,
+            "output_selection": None,
+            "keep_l_in_keys": False,
+            "arrays_backend": None,
+            "cg_backend": None,
+        }
+        self._density_correlations_hypers = density_correlations_hypers
+        self._density_correlations_calc = DensityCorrelations(
+            **density_correlations_hypers
+        )
+
+    def forward(self, frames) -> TensorMap:
+        """
+        Calls the :py:meth:`LambdaSoap.compute` function.
+
+        This is intended for :py:class:`torch.nn.Module` compatibility, and should be
+        ignored in pure Python mode.
+        """
+        return self.compute(frames)
+
+    def compute(self, frames) -> TensorMap:
+        """
+        Computes the Lambda-SOAP descriptor for the given ``frames``.
+
+        :param frames: The input frames, i.e. ase Atoms objects.
+        """
+        # Compute the nu = 1 density and move "neighbor_type" to properties
+        density = self._spherical_expansion_calc.compute(frames)
+        if self._atom_types is None:
+            keys_to_move = ["neighbor_type"]
+        else:
+            keys_to_move = Labels(
+                names=["neighbor_type"],
+                values=_dispatch.int_array_like(
+                    self._atom_types, like=density.keys.values
+                ).reshape(-1, 1),
+            )
+        density = density.keys_to_properties(keys_to_move)
+
+        # Compute the nu = 2 lambda-SOAP descriptor
+        lsoap = self._density_correlations_calc.compute(density)
+
+        return lsoap
+
+
+class LambdaSoapXSphExByPair(TorchModule):
```

Comment:
This is `\rho^2 x g_ij`, right? since `g_ij` is `SphericalExpansionByPair`, this should be `BispectrumByPair` IMO (with potential `Lambda`, `Soap`, `Equivariant` in there as well).